### ISSUE TITLE
[$2000] Newline is removed from code block with a character before it

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#16982]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Send this message to a chat
|
```
test
```
Case 1:
 1. Copy the whole message by ctrl/cmd+c
 2. Paste it to the composer
 3. Notice the ``` is at the same line as |
This only happens with | character
Case 2:
 1. Edit the message
 2. Notice the ``` is at the same line as |
This happens to all character

## Expected Result:
` ``` `should be on its own line

## Actual Result:
` is at the same line as | because the newline is removed

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [ ] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.2.95-0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/43996225/230213844-6ba0a016-56cb-42f6-b275-d587211fa6b4.mov

https://user-images.githubusercontent.com/43996225/230213878-fa518834-5fcd-43df-9ff5-539214d4c9cb.mp4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @bernhardoj 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1680669746873359

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01a8161082bc95b676</li>
        <li>Upwork Job ID: 1643736935146340352</li>
        <li>Last Price Increase: 2023-05-02</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @twisterdotcom (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01a8161082bc95b676

**Comment by MelvinBot** [comment]
Current assignee @twisterdotcom is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @0xmiroslav (`External`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @PauloGasparSv (`External`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by ImBIOS** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The ``` is at the same line as | when we write it in a different line. We wanted to make it still in their own line.

### What is the root cause of that problem?

In Markdown, the | symbol is used to separate columns in a table, and it is not intended to have a line break when it is copied and pasted. When you copy and paste a table created in Markdown, the | symbol is used to indicate the boundaries between columns, and it is preserved in the copied text.

### What changes do you think we should make in order to solve the problem?


![image](https://user-images.githubusercontent.com/41441643/230240232-a6275e65-061c-460c-afb8-95fb63489730.png)

We need to manually check if this symbol exists and add a `<br />` in HTML and `\n` in text format.

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief, and avoid jargon. Feel free to use images, charts, or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

There are several libraries out there to handle this problem, but it requires more changes in the code and I didn't recommend this method.

**Comment by JMGCode** [comment]
Please re-state the problem that we are trying to solve in this issue.
The ``` is at the same line as | when we write it in a different line. We wanted to make it still in their own line.

What is the root cause of that problem?
It's a bad regex on the ExpensiMark replaceBlockWithNewLine function

What changes do you think we should make in order to solve the problem?
change regex

Solution video
https://drive.google.com/file/d/1endHanNM9x47XcqpmRXvB-SgS5l4Kw63/view?usp=share_link

**Comment by MelvinBot** [comment]
📣 @JMGCode! 📣

Hey, it seems we don’t have your contributor details yet! You'll only have to do this once, and this is how we'll hire you on Upwork.
Please follow these steps:

1. Get the email address used to login to your Expensify account. If you don't already have an Expensify account, create one [here](https://new.expensify.com/). If you have multiple accounts (e.g. one for testing), please use your main account email.
2. Get the link to your Upwork profile. It's necessary because we only pay via Upwork. You can access it by logging in, and then clicking on your name. It'll look like [this](https://www.upwork.com/freelancers/~01ca8ad8ce36b9eceb). If you don't already have an account, sign up for one [here](https://www.upwork.com/nx/signup/?dest=home).
3. Copy the format below and paste it in a comment on this issue. Replace the placeholder text with your actual details.

![Screen Shot 2022-11-16 at 4 42 54 PM](https://user-images.githubusercontent.com/30609178/202326034-1a292e64-48fd-4fef-9f4f-7feaaeedfd9f.png)

Format:
```
Contributor details
Your Expensify account email: <REPLACE EMAIL HERE>
Upwork Profile Link: <REPLACE LINK HERE>
```

**Comment by JMGCode** [comment]
Contributor details
Your Expensify account email: jmgcode@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01471144f7a600d0ea

**Comment by MelvinBot** [comment]
✅ Contributor details stored successfully. Thank you for contributing to Expensify!

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Two issues need to solve

1. Copy and paste code block message wth pipe character before misses a line break
2. '```' isn't displayed in new line in editing mode

### What is the root cause of that problem?

From Web(Chrome/Safari), when we copy message through ctrl/cmd+c, we set html selection to clipboard [here](https://github.com/Expensify/App/blob/78a9b540f6244276599d58f9aa3fc7d671441c8d/src/components/CopySelectionHelper.js#L40). 

We send following two new comments to dig the root cause

````
|
```
code
```
````

````
A
```
code
```
````

like 

<img width="313" alt="image" src="https://user-images.githubusercontent.com/117511920/230256251-efb0abbf-ab28-40c0-8ec6-2a7a58d6a296.png">

We add a log before this line and copy the above message through ctrl/cmd+c. The selection html is shown as below

<img width="693" alt="image" src="https://user-images.githubusercontent.com/117511920/230256895-aa400a54-05e1-43d2-9686-fe6902312ebb.png">

If we paste in composer, then we convert html to markdown [here](https://github.com/Expensify/App/blob/78a9b540f6244276599d58f9aa3fc7d671441c8d/src/components/Composer/index.js#L216)

HTML selection 

```html
<comment><div>| </div><pre><span>code</span></pre></comment>
```

is converted into markdown

````md
| ```
code
```
````

While HTML selection 

```html
<comment><div>A </div><pre><span>code</span></pre></comment>
```

is converted into markdown

````md
A 
```
code
```
````

To figure out why the pipe `|` character case behaves differently, we can have a closer look at [htmlToMarkdown](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L503-L525) method. The `div` tag(replaced with `[block]`) should also be translated into a line break for pipe `|` character case in [replaceBlockWithNewLine](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L468). But there's a flaw in [this regex](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L486) that it can match pipe `|` character. So `text.match(/[\n|>][>]?[\s]?$/)` returns true and skips to add a line break. See regex testing picture below

<img width="600" alt="image" src="https://user-images.githubusercontent.com/117511920/230260000-5abbbf60-f22e-421d-a7ab-e2dab3c9c96d.png">

The above is the RCA of copy and paste issue for pipe `|` character case.

Let's move forward to dig the line break issue in editing mode.

As the line break missing issue in editing mode is same for having pipe `|` character case, let's send following message to dig the root cause

````
A
```
code
```
````

We translate markdown text into html [here](https://github.com/Expensify/App/blob/78a9b540f6244276599d58f9aa3fc7d671441c8d/src/libs/ReportUtils.js#L909) and save following message html to backend. 

```html
A <pre>code<br /></pre>
```

When replacing markdown to html, we remove the line break after `A` and replace it with a space ' ' using this [rule](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L186-L191). That's why there's an extra space after `A` in the message html.

When editing a sent message, we call [htmlToMarkdown](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L503) to convert message html to markdown text [here](https://github.com/Expensify/App/blob/78a9b540f6244276599d58f9aa3fc7d671441c8d/src/pages/home/report/ReportActionItemMessageEdit.js#L91). The message html of above sent comment is 

```html
A <pre>code<br /></pre>
```

which is much simpler than the html actually rendered in the page.

In method [htmlToMarkdown](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L503), the above message html is translated into following markdown text by applying [newline rule](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L217-L230) and [codeFence rule](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L271-L275).

````
A ```
code
```
````

The root cause is the way we translate codefence from markdown-to-html and from html-to-markdown.

When digging the root cause, I went through this closed issue https://github.com/Expensify/App/issues/12783 which explains the root cause and ends up do nothing.

### What changes do you think we should make in order to solve the problem?

To fix the first issue for pipe `|` character case, we can remove the unnecessary pipe `|` character in this [regex](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L486). 

I tried to come up a solution for the second issue, but can't find a perfect one that could avoid introducing new regressions.

I agree with @[iwiznia](https://github.com/iwiznia) to save the original text entered by user to solve this issue as he commented [here](https://github.com/Expensify/App/issues/12783#issuecomment-1322053652).

### What alternative solutions did you explore? (Optional)

N/A

**Comment by JMGCode** [comment]
https://drive.google.com/file/d/1_9Lf-JX49tW4FE1NR8KVoVL7mN8KJa19/view?usp=share_link

**Comment by PauloGasparSv** [comment]
> @PauloGasparSv @twisterdotcom should we also consider Case 2 as bug and fix it here?

I don't think we should consider fixing either anymore!

I just checked all proposals and @eh2077 found an [older issue](https://github.com/Expensify/App/issues/12783) from 2022 where we already discussed Case 2. It links to a [discussion on slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1669040394837259) we had on both bugs (it also mentions other discussions we had on those) and back then we decided to **do nothing!** here:
> Asked around to more people and we decided that:
> so long as the edited version is semantically the same as the original (eg, both produce the same HTML), it's not a bug.
> So, with that, going to close the issues I linked above.

Because of that I think we should **close the issue and not fix anything**, ok @twisterdotcom?
(Also, I'm not sure if @eh2077 proposal/investigation should be compensated since **it helped decide what to do here**)

**Comment by eh2077** [comment]
@PauloGasparSv Thanks for reviewing proposals!

But I think we should at least [this regex](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L486) demonstrated by **case 1** as the pipe `|` character is not needed in the regex.

**Comment by PauloGasparSv** [comment]
> @PauloGasparSv makes sense, but that was old discussion. Can you confirm again on slack before close?

Sure thing @0xmiroslav, will create a new thread in expensify-open-source (and link the previous one + this issue + cc you guys) so we can involve more people in the discussion and propose that we only fix Case 1.

I just need some time to read all the previous threads and make sure I'm not missing anything or mentioning something we already discussed, so I'll create it after lunch Today!

**Comment by esh-g** [comment]
@bernhardoj I think you meant to tag @eh2077 this happened to me the second time today 😂

**Comment by PauloGasparSv** [comment]
Will have to start the slack discussion tomorrow, couldn't find time for it Today!

**Comment by PauloGasparSv** [comment]
Same as yesterday, could not start the discussion because I was too focused on some [N6.99](https://github.com/Expensify/Web-Expensify/pull/36987) [N7](https://github.com/Expensify/App/pull/16962) P.R.'s and a [deploy blocker](https://github.com/Expensify/App/pull/17428).

Will try again to start the discussion tomorrow.

**Comment by MelvinBot** [comment]
@PauloGasparSv, @twisterdotcom, @0xmiroslav Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by PauloGasparSv** [comment]
Sorry for the delay on this, still couldn't find time to start the thread. Now I'm planning on starting it before Wednesday and If that doesn't work I'll find someone to take this over.

**Comment by twisterdotcom** [comment]
Hey @PauloGasparSv - I'm back from OOO now. Is there anything I can help with here?

**Comment by PauloGasparSv** [comment]
Hey @twisterdotcom!

**Context**
We had some discussions about this issue in November of last year and we decided to do nothing and not fix this back then (check [this comment](https://github.com/Expensify/App/issues/16982#issuecomment-1505271405)).

This problem [happens in 2 different places for 2 different reasons](https://github.com/Expensify/App/issues/16982#issuecomment-1501216011):
1. Copy/Paste is one issue
2. Edit message is another issue

Despite the fact that we already discussed this error before, we have some proposals to fix the first copy/paste case. So I was going to ask the following on slack but I think it makes more sense if you give your opinion first!

**Question**
Should we do nothing like last time or only fix case 1?

**Comment by twisterdotcom** [comment]
I'm actually inclined to agree with @0xmiroslav here: https://github.com/Expensify/App/issues/16982#issuecomment-1505367747.

I think we should aim for consistency. I sort of agree that:
>so long as the edited version is semantically the same as the original (eg, both produce the same HTML), it's not a bug.

But I also think there is an expectation that what you would edit would match what you just typed. This is obviously super small, but the fix itself is also super small, so let's just do it.

**Comment by PauloGasparSv** [comment]
>  This is obviously super small, but the fix itself is also super small, so let's just do it.

Ok, sorry for the delay here then team!

Let's **try to fix both cases** then and if that's not possible then **at least fix case 1.** @0xmiroslav should we wait for proposal updates/more proposals?

**Comment by PauloGasparSv** [comment]
Hey team, I'm going to QA https://github.com/Expensify/Web-Expensify/pull/37040 using this issue so I'll quickly double the title's compensation then change it back to $1000.

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $2000

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1250

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $1000

**Comment by mountiny** [comment]
<img width="510" alt="image" src="https://user-images.githubusercontent.com/36083550/232906004-f0f534e9-66ef-444e-8d20-6b5832d595d9.png">
worked

**Comment by MelvinBot** [comment]
@PauloGasparSv @twisterdotcom @0xmiroslav this issue was created 2 weeks ago. Are we close to approving a proposal? If not, what's blocking us from getting this issue assigned? Don't hesitate to create a thread in #expensify-open-source to align faster in real time. Thanks!

**Comment by MelvinBot** [comment]
📣 It's been a week! Do we have any satisfactory proposals yet? Do we need to adjust the bounty for this issue? 💸

**Comment by twisterdotcom** [comment]
What do you think? Increase the price then?

**Comment by MelvinBot** [comment]
Upwork job price has been updated to $2000

**Comment by eh2077** [comment]
I'd like to share what I found about [this rule](https://github.com/Expensify/expensify-common/blob/a49e4b38d412669a7c91b79fab82c7f3a9fc4fe7/lib/ExpensiMark.js#L186-L191) to replace the line break before `<pre>` tag
```js
            {
                // We're removing <br /> because when <br /> and <pre> occur together, an extra line is added.
                name: 'replacebr',
                regex: /<br\s*[/]?><pre>\s*/gi,
                replacement: ' <pre>',
            },
```
Why we should add a space before `<pre>` tags after replacing? I found that this matters if there's an empty line above the codeblock, like for input
````txt
A

```code```
````
We need to display the empty line before codeblock and it won't work if not adding the extra space.

I inspected the browser HTML and below picture shows the difference of HTML with and without a space added 

<img width="800" alt="image" src="https://user-images.githubusercontent.com/117511920/232938284-eec0c9c9-1f2b-404a-b27b-f568c7bbce1f.png">

I'm not sure if this is related to how the html render renders the `<br />` before block emelent `<pre>` as we're using customer html render in the App codebase.

**Comment by MelvinBot** [comment]
@PauloGasparSv @twisterdotcom @0xmiroslav this issue is now 3 weeks old. There is one more week left before this issue <a href='https://stackoverflowteams.com/c/expensify/questions/15233/15234#15234'>breaks WAQ</a> and will need to go internal. What needs to happen to get a PR in review this week? Please create a thread in #expensify-open-source to discuss. Thanks!

**Comment by twisterdotcom** [comment]
@PauloGasparSv or @0xmiroslav any insight here?

**Comment by MelvinBot** [comment]
@PauloGasparSv @twisterdotcom @0xmiroslav this issue is now 4 weeks old and preventing us from maintaining WAQ, can you:
- Decide whether any proposals currently meet our guidelines and can be approved as-is today
- If no proposals meet that standard, please take this issue internal and treat it as one of your highest priorities
- If you have any questions, don't hesitate to start a discussion in #expensify-open-source

Thanks!

**Comment by MelvinBot** [comment]
Current assignee @0xmiroslav is eligible for the Internal assigner, not assigning anyone new.

**Comment by cead22** [comment]
Adding my 2 cents here.
- First I was going to suggest exactly what Paulo suggested in [this comment](https://github.com/Expensify/App/issues/16982#issuecomment-1505271405) and based on the conclusion from that old discussion that _so long as the edited version is semantically the same as the original (eg, both produce the same HTML), it's not a bug._
- I then read Ted's [comment](https://github.com/Expensify/App/issues/16982#issuecomment-1513025047) saying _This is obviously super small, but the fix itself is also super small, so let's just do it._
- But then we started talking about how the solution is not super small on simple and can cause regressions, and how this is complex enough to increase the price of the issue

It seems to me like the solution is not super small, and thus, I think it's worth closing this issue.

**Comment by twisterdotcom** [comment]
@0xmiroslav I discussed this with @cead22 internally and I'm inclined to agree now again that as long as the HTML is the same, we can count this as not a bug. If you do seriously think there is a direction we can go, please comment as such.

I can't close without saying thanks for the contributions though to @ahmdshrif and @eh2077 especially.

**Comment by eh2077** [comment]
@twisterdotcom I think we should fix the first case about the pipe `|` character.

Hmm, do I have luck to get some compensation for investigating the issue as @PauloGasparSv mentioned in comment https://github.com/Expensify/App/issues/16982#issuecomment-1505271405 ?

**Comment by twisterdotcom** [comment]
I've sent an offer of $250 as a bonus for the investigation @eh2077. Thanks for your work on this here.

**Comment by eh2077** [comment]
@twisterdotcom thanks for the offer! Accepted.

**Comment by cead22** [comment]
re: the **in general** section, it seems like we're still talking about hypotheticals, so it's hard to engage with those points without having concrete examples

re: the video you're showing, perhaps that's where the disconnect is. IMO, we care that the resulting HTML is the same **in new dot**, not in any other app.

![](http://g.recordit.co/c85IBey63s.gif)

**Comment by cead22** [comment]
> @cead22 What do you think about the opposite case? when copying this case from another resource to a new dot do we care about it?
> What is the reason for using different Markdown rules?

Can you ask these in #expensify-open-source?

**Comment by twisterdotcom** [comment]
Can we wait on the outcome of the investigation in https://github.com/Expensify/App/issues/18928? I don't think we should double up work here.

**Comment by nikosamofa** [comment]
@0xmiroslav,
The [proposal](https://github.com/Expensify/App/issues/26490#issuecomment-1758233329) for #26490  fixes this issue
Is it okay to create a PR? If so, what issue can I reference? #16982 or #26490?

https://github.com/Expensify/App/assets/30532000/7cd8c0b8-b81a-4ba0-9840-c5c2608d1840

**Comment by cead22** [comment]
26490
